### DISCUSSION TITLE
How to add scope permanently?

### DISCUSSION THREAD

**Discussion by ariel-frischer** [Zhengqbbb/cz-git#104] [General]
When I add a custom scope once, I want it to be in the list of available scopes the next time I run the cli tool. Is there a setting for this? Do I have to add ALL possible scopes to my config file beforehand?

**Comment by Zhengqbbb** [comment]
I'm sorry, I'm afraid that's not available. I don't have the data storage capabilities of your repo. 🤔

**Comment by ariel-frischer** [comment]
Aw ok

**Comment by Zhengqbbb** [comment]
I just think if I can use [formatMessageCB](https://cz-git.qbb.sh/config/engineer#formatmessagecb), which is a callback Fn, and I'm not sure if it can write in the logic of the scope and call the Node API to write the configuration (.git /scope.json)

But this is just my idea, I'm not sure if it's possible, need to try it and test, lol 🧐

**Reply by ariel-frischer** [reply]
Yes, I think its a great feature :)

**Comment by Zhengqbbb** [comment]
@ariel-frischer Hi. The demo code is ready. I think it can resolve your request
Source code. See: https://github.com/Zhengqbbb/czgit-playground/tree/cache-scope
Thx your support ~

<br>

https://user-images.githubusercontent.com/40693636/226102118-8f76e0a4-3f2c-410f-a451-c4bbe6d5cedb.mp4

**Comment by ariel-frischer** [comment]
Wow its working perfectly very impressive I also like how the scope is saved to each .git folder, so every project can have custom scopes. Will you be merging these changes into your main branch? For anyone wondering how to use this,  I did a git clone, checked out the correct branch with `git checkout cache-scope`. Then I run `npm i`. Your cli binaries are inside `/node_modules/.bin/` on my macbook.

**Reply by Zhengqbbb** [reply]
> these changes into your main branch?

I'm afraid not.
These are all features done under the configuration, all user-defined, the user decides how much to enhance, that's the power of scripting
I'll put it in the recipes on the website afterwards
